First pass at updating secrets management and ingress

timfallmk opened the pull request:
This should bring secrets storage more in line with best practices, and also cut down on the errors surrounding encoding and loading. Also a first step to eventually doing away with storing secrets on raw volumes, which is no longer recommended practice.

`scimsession` handling
Loading of the scimsession is now handled more directly and can be done in one of two ways:

1. Passed in as a single raw `json` string with no trailing newline. This can be done in a `values.yaml` file or via `--set` at runtime. Should work nicely with CI/CD pipelines.
2. Using `--set-file=scim.credentialsSecret.data=<path to file on local disk>` to avoid anything appearing in logs, and to go straight from the download screen during SCIM setup on 1Password.com
3. (Bonus). For extra brownie points you can use the 1Password cli to reference the `scimsession` file stored in a 1Password vault and deference that secret at runtime

Much delayed, but as promised https://github.com/1Password/op-scim-helm/issues/53#issuecomment-1091100056

This is only a first pass at cleanup, and mostly to address the biggest pain point of `scimsession` loading. There are more fundamental refactors to be done around scoping and how the SCIM Bridge handles external TLS forwarding, ideally all with an eye to bringing much of the logic around TLS, URL assignment, ports, etc. out of the app code and into the deployment.

timfallmk commented on op-scim-bridge/values.yaml:
```diff
@@ -46,16 +49,12 @@ scim:
+    className:
+    domain:
+    path: \
+    tls:
```
Note that this makes https://github.com/1Password/op-scim-helm/pull/60/files#diff-0de2f4f1369c8e7f265bf4aa8b3996820bfe58482ec75ee491212a670e29cbedR80-R82 redundant but I wasn't sure how you wanted to handle that as it may break things on your end.

grellyd replied:
I'm afraid this link does not resolve anywhere obvious. I assume this is because it is outside the diff. Am I correct in thinking you are referring to `values.yaml` lines 80-82?

AKA

```go
tls:
    enabled: false
    secret: "{{ .Chart.Name }}-bridge-tls"
 ```

nhuray commented:
@timfallmk I would like to deploy the bridge with that Helm chart using `credentialSecrets`.

But, at the moment I can't disable `credentialVolume` so I'm waiting for that PR to be merged. 

Is there anything I can do to help ?

nhuray commented on op-scim-bridge/templates/service.yaml:
```diff
@@ -14,7 +14,11 @@ metadata:
     {{- toYaml . | nindent 4 }}
   {{- end }}
 spec:
+  {{- if not .Values.scim.ingress.enable }}
```
spot a typo here: `.Values.scim.ingress.enable` -> `.Values.scim.ingress.enabled`

nhuray replied:
spot a typo here: `.Values.scim.ingress.enable` -> `.Values.scim.ingress.enabled`

timfallmk commented:
@nhuray  You could use the branch in this fork right away if you wanted to. There's only one major thing blocking this from working, and it's something I can't do anything about.

The `OP_SESSION` env variable expects a file _path_ and not the file contents itself. As it currently is, credentials secrets never would have worked to begin with as they don't place files on disk. The code needs to be changed to take in the contents of the file itself, not just a path. Since the code for the bridge itself is not public, I can't fix it.

@nhuray I would recommend pinging some of the maintainers to see if you can get a response.

grellyd commented on op-scim-bridge/templates/secret.yaml:
```diff
@@ -18,12 +20,7 @@ metadata:
   {{- end }}
 type: Opaque
-stringData:
+data:
   {{ .Values.scim.credentialsSecret.key }}: |-
-  {{ if .Values.scim.credentialsSecret.value_json }}
-  {{- .Values.scim.credentialsSecret.value_json | b64enc | indent 2 }}
-  {{ else }}
-  {{- .Values.scim.credentialsSecret.value_base64 | indent 2 }}
-  {{ end }}
-{{- end }}
```
We would like to maintain the ability to ingest both json and raw base64, as they each have a place in differing workflows. However what we can do is rename `value_json` to just be `value` or something similarly default. JSON ingestion is the primary use case. 

This would replicate the [credentials ingestion for the 1Password Connect server](https://github.com/1Password/connect-helm-charts/blob/main/charts/connect/templates/connect-credentials.yaml) which has both `connect.credentials` and `connect.credentials_base64`

timfallmk replied:
Unfortunately that's an anti-pattern for Helm which is why you won't find the mechanism built in.

However I think you're confusing what happens. The 'json' string can be passed in, as I said in my PR notes, and the k8s API will automatically encode it as a bse64 string for storage in a secret (part of the reason to use secrets, instead of volumes). Taking it in as a file does the same thing, expect this time reading the string from file contents instead of `stdin`.

If you really want to have the ability for people to encode the string themselves (not sure why you would want to do this when it happens anyway, and people are just prone to insert errors in their encoding), then I suggest you make that a mechanism outside this loop, as it breaks a number of core Helm functionality assumptions.

jlane67 replied:
This is true.  JSON, PEM-formatted certificates, whatever, k8s will store it as a base64 single line string if you create a secret.  The only thing adding a manual "pre-base64'd" option does is create possible breakage.  Even if you're concerned about writing a migration tool for existing users, reading a file out of a k8s `volume` and writing it to a `secret` can be handled automatically, and if the file is already base64'd, just undo that before calling `create secret`.

grellyd commented on op-scim-bridge/values.yaml:
```diff
@@ -6,10 +6,13 @@ scim:
   # credentialsVolume configures the SCIM bridge to access the credentials via a volume
   credentialsVolume:
+    # enables the `credentialsVolume`. Defaults to `false`. Either `credentialsSecret` or `cedentialsVolume` must be specified, but not both
+    enabled: false
```
This helm chart is the base upon which our marketplace applications are built. As they do the 'one-click' installation method which writes the value to the volume, we will need `credentialsVolume` to be default and `credentialsSecret` to be the alternate. 

For customers wishing to leverage the credentials secret directly, swapping the enabled flags themselves before deployment should be doable.

timfallmk replied:
These can be changed, but I would highly encourage you to change the marketplace entries to follow the secrets best practices. At the very least it would be helpful to call out that people should enable using secrets in the docs somewhere when they do the deploy themselves.

grellyd commented on op-scim-bridge/values.yaml:
```diff
@@ -20,15 +23,15 @@ scim:
+    name: "{{ .Chart.Name }}-bridge-credentials"
     # key of the secret
-    # key: scimsession
-    # value_json is the JSON contents of the scimsession file
-    # value_json: "{}"
-    # value_base64 is the base64 encoded contents of the scimsession file
-    # value_base64: "base64 encoded scimsession file"
```
Unfortunately we will need to keep these as we wish to maintain support for both JSON and Base64 encoded scimsessions.

timfallmk replied:
See my above comments

grellyd commented on op-scim-bridge/templates/secret.yaml:
```diff
@@ -1,7 +1,9 @@
-{{- if or .Values.scim.credentialsSecret.value_json .Values.scim.credentialsSecret.value_base64 }}
-{{- if and .Values.scim.credentialsSecret.value_json .Values.scim.credentialsSecret.value_base64 }}
-  {{- fail "Only one of scim.credentialsSecret.value_json and scim.credentialsSecret.value_base64 can be specified" }}
+{{- if .Values.scim.credentialsSecret.enabled }}
+{{- if and .Values.scim.credentialsSecret.enabled .Values.scim.credentialsVolume.enabled }}
+  {{- fail "Either credentialsSecret or credentialsVolume may be specified, not both"}}
+{{- end }}
```
Fantastic. Thanks for this change. 🎉  I believe our exiting check may not have worked as desired.

grellyd commented on op-scim-bridge/templates/deployment.yaml:
```diff
@@ -98,10 +98,10 @@ spec:
             - name: "OP_SESSION"
-              {{- if .Values.scim.credentialsVolume }}
+              {{- if .Values.scim.credentialsVolume.enabled }}
               value: "/home/opuser/.op/{{ .Values.scim.credentialsVolume.file }}"
               {{- end }}
-              {{- if .Values.scim.credentialsSecret }}
+              {{- if .Values.scim.credentialsSecret.enabled }}
```
This is an absolute improvement over the commenting out of sections. Thank you! 🙏

pniederlag replied:
please provide separate toggles for managing the secret and for injecting the secret ... see #87 for details

grellyd commented on op-scim-bridge/templates/secret.yaml:
```diff
@@ -1,7 +1,9 @@
+{{- if and .Values.scim.credentialsSecret.enabled .Values.scim.credentialsVolume.enabled }}
+  {{- fail "Either credentialsSecret or credentialsVolume may be specified, not both"}}
+{{- end }}
+{{- if not .Values.scim.credentialsSecret.data }}
+  {{- fail "Contents of the scimsession file must be provided" }}
```
Given we want to keep both data options, this will probably have to look something like:

```yaml
{{- if and (.Values.scim.credentialsSecret.data) (.Values.scim.credentialsSecret.data_base64) -}}
  {{- fail "Only one of scim.credentialsSecret.data and cim.credentialsSecret.data_base64 can be specified" -}}
{{- end -}}

timfallmk replied:
I would recommend this be brought out of this logic loop (maybe a helper template function) since users of helm would expect it to function a certain way, which is counter to how this works. I can understand the need to keep backward compatibility, but you might be able to do both by moving it.

grellyd commented on op-scim-bridge/templates/ingress.yaml:
```diff
@@ -60,5 +58,6 @@ spec:
           {{- end }}
-    {{- end }}
+    {{ else }}
+      {{- fail "A domain for the deployment must be specified. This is *not* a 1Password.com address. Example: '1password-scim-bridge.yoursite.com'" }}
```
Great change. Thank you.

grellyd commented on op-scim-bridge/templates/ingress.yaml:
```diff
@@ -29,25 +29,23 @@ spec:
-        {{- range .hosts }}
-        - {{ . | quote }}
-        {{- end }}
-      secretName: {{ .secretName }}
-    {{- end }}
+    - hosts: 
+        - {{ .domain | quote }}
+      # TODO: Figure out why templating this in the values file doesn't work
+      secretName: "{{ $.Chart.Name }}-bridge-tls"
+  {{- end }}
   {{- end }}
   rules:
-    {{- range .Values.scim.ingress.hosts }}
-    - host: {{ .host | quote }}
+    {{- if .Values.scim.ingress.domain }}
+    - host: {{ .Values.scim.ingress.domain | quote }}
       http:
         paths:
-          {{- range .paths }}
-          - path: {{ .path }}
-            {{- if and .pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
-            pathType: {{ .pathType }}
```
These are thought provoking changes.

Can you speak a little more to your intention around these changes? Why strip out the range of hosts?

The range was introduced by another PR which at the time convinced us of the utility.

timfallmk replied:
The spec actually only accepts a single entry https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#ingressrule-v1-networking-k8s-io

I think what the previous one was going for was a difference in path based vs. host based routing. You can do either, but not both. Keeping the range in makes enumerating the sub-keys a lot harder though. Just my 2 cents.

timfallmk replied:
I should clarify:
I know a lot of helm charts get a bit "lazy" and just stub out the entire ingress template, expecting the user to provide their own. Personally I find that a little annoying, but for new users especially it is helpful to simply provide keys (with defaults) for the major required ingress fields and construct it ourselves.

grellyd reviewed (changes requested):
@timfallmk Thank you very much for these changes! Your PR is very appreciated.

grellyd commented:
> 

Indeed you may use this PR as is if you wish @nhuray. 

Regarding the scimsession ingestion, indeed the raw value to `OP_SESSION` can be a path to a file. However it may also be a base64 encoded scimsession file. That should allow you to directly ingest the contents of the scimsession without providing a path. 

Let me know how that works.

timfallmk commented:
Unfortunately that doesn’t seem to be the case. You can run this PR as is and test it for yourself; that’s how it’s currently set up.
  

  
The reason I know this is we’re in the process of signing a large PO with agile bits and this was a show stopper for us. Our account rep,   ***@***.*** confirmed it was a bug with a fix in the works.
  
  
  
  
  
  
  

  
  
>   
> On May 27, 2022 at 15:54,  <Graham ***@***.***)>  wrote:
>   
>   
>   
>
>
>
>   
>
>   
>
> Indeed you may use this PR as is if you wish  @nhuray (https://github.com/nhuray).
>
>   
>
> Regarding the scimsession ingestion, indeed the raw value to  OP_SESSION  can be a path to a file. However it may also be a base64 encoded scimsession file. That should allow you to directly ingest the contents of the scimsession without providing a path.
>
>   
>
> Let me know how that works.
>
>   
>
> —
>  Reply to this email directly,   view it on GitHub (https://github.com/1Password/op-scim-helm/pull/60#issuecomment-1140089657), or   unsubscribe (https://github.com/notifications/unsubscribe-auth/ABIR4FMD5U4NFYDZGIX2ZR3VMFHB7ANCNFSM5WDUL6MA).
>  You are receiving this because you were mentioned.Message ID:  ***@***.***>
>
>   
>

timfallmk commented:
https://github.com/1Password/op-scim-helm/pull/60#discussion_r884014758 Not sure what question you were asking here, can you clarify?
There's also a typo there as the path should be `/`. I'll fix that.

timfallmk commented:
This seems to have gotten stale. Should I close it?

devillexio commented:
Hi @timfallmk.

Regarding: 
> This seems to have gotten stale. Should I close it?

Yes please. I'd appreciate if we can raise issues for any of the outstanding items that should be addressed.

timfallmk commented:
We've fixed the bugs for our purposes by maintaining our own fork, as maintainers seem uninterested in making substantive corrections or improvements.

If that ever changes this can be reopened and rebased.
